### PR TITLE
`Search` - update API version to `2024-06-01-preview`

### PR DESCRIPTION
**Pull request by jan-mrm**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description

This PR updates all `Search` service api versions to `2024-06-01-preview`.
This is done as a prerequisite for issue #26575

Notable change is that the spec now returns a `identity.SystemAndUserAssignedMap` instead of a `identity.SystemAssigned` allowing for more than just `SystemAssigned` identities.

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.

Affected resources from the api update:
- azurerm_search_service (resource)
- azurerm_search_service (data source)
- azurerm_search_shared_private_link_service
- azurerm_cognitive_account

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

```
=== RUN   TestAccSearchService_basicSku
=== PAUSE TestAccSearchService_basicSku
=== RUN   TestAccSearchService_standardSku
=== PAUSE TestAccSearchService_standardSku
=== RUN   TestAccSearchService_requiresImport
=== PAUSE TestAccSearchService_requiresImport
=== RUN   TestAccSearchService_complete
=== PAUSE TestAccSearchService_complete
=== RUN   TestAccSearchService_update
=== PAUSE TestAccSearchService_update
=== RUN   TestAccSearchService_ipRules
=== PAUSE TestAccSearchService_ipRules
=== RUN   TestAccSearchService_identity
=== PAUSE TestAccSearchService_identity
=== RUN   TestAccSearchService_identityUserAssigned
=== PAUSE TestAccSearchService_identityUserAssigned
=== RUN   TestAccSearchService_identitySystemAndUserAssigned
=== PAUSE TestAccSearchService_identitySystemAndUserAssigned
=== RUN   TestAccSearchService_hostingMode
=== PAUSE TestAccSearchService_hostingMode
=== RUN   TestAccSearchService_hostingModeInvalidSKU
=== PAUSE TestAccSearchService_hostingModeInvalidSKU
=== RUN   TestAccSearchService_partitionCountInvalidBySku
=== PAUSE TestAccSearchService_partitionCountInvalidBySku
=== RUN   TestAccSearchService_partitionCountInvalidByInput
=== PAUSE TestAccSearchService_partitionCountInvalidByInput
=== RUN   TestAccSearchService_replicaCountInvalid
=== PAUSE TestAccSearchService_replicaCountInvalid
=== RUN   TestAccSearchService_replicaCount
=== PAUSE TestAccSearchService_replicaCount
=== RUN   TestAccSearchService_customerManagedKeyEnforcement
=== PAUSE TestAccSearchService_customerManagedKeyEnforcement
=== RUN   TestAccSearchService_customerManagedKeyEnforcementUpdate
=== PAUSE TestAccSearchService_customerManagedKeyEnforcementUpdate
=== RUN   TestAccSearchService_apiAccessControlRbacError
=== PAUSE TestAccSearchService_apiAccessControlRbacError
=== RUN   TestAccSearchService_apiAccessControlUpdate
=== PAUSE TestAccSearchService_apiAccessControlUpdate
=== RUN   TestAccSearchService_localAuthEnabled
=== PAUSE TestAccSearchService_localAuthEnabled
=== CONT  TestAccSearchService_basicSku
=== CONT  TestAccSearchService_hostingMode
=== CONT  TestAccSearchService_requiresImport
=== CONT  TestAccSearchService_standardSku
=== CONT  TestAccSearchService_customerManagedKeyEnforcement
=== NAME  TestAccSearchService_semanticSearchBasicFreeSkuError
=== CONT  TestAccSearchService_apiAccessControlRbacError
=== CONT  TestAccSearchService_customerManagedKeyEnforcementUpdate
--- PASS: TestAccSearchService_apiAccessControlRbacError (32.42s)
--- PASS: TestAccSearchService_requiresImport (81.91s)
=== CONT  TestAccSearchService_partitionCountInvalidByInput
--- PASS: TestAccSearchService_partitionCountInvalidByInput (4.01s)
=== CONT  TestAccSearchService_replicaCount
--- PASS: TestAccSearchService_standardSku (91.53s)
=== CONT  TestAccSearchService_replicaCountInvalid
--- PASS: TestAccSearchService_basicSku (99.56s)
=== CONT  TestAccSearchService_partitionCountInvalidBySku
--- PASS: TestAccSearchService_replicaCountInvalid (32.43s)
=== CONT  TestAccSearchService_hostingModeInvalidSKU
--- PASS: TestAccSearchService_partitionCountInvalidBySku (32.31s)
=== CONT  TestAccSearchService_identity
=== CONT  TestAccSearchService_identitySystemAndUserAssigned
--- PASS: TestAccSearchService_hostingModeInvalidSKU (32.89s)
--- PASS: TestAccSearchService_identitySystemAndUserAssigned (168.33s)
=== CONT  TestAccSearchService_identityUserAssigned
--- PASS: TestAccSearchService_customerManagedKeyEnforcement (619.92s)
=== CONT  TestAccSearchService_ipRules
--- PASS: TestAccSearchService_hostingMode (670.34s)
=== CONT  TestAccSearchService_update
--- PASS: TestAccSearchService_localAuthEnabled (667.52s)
=== CONT  TestAccSearchService_complete
--- PASS: TestAccSearchService_customerManagedKeyEnforcementUpdate (701.05s)
--- PASS: TestAccSearchService_identity (672.80s)
--- PASS: TestAccSearchService_apiAccessControlUpdate (792.87s)
--- PASS: TestAccSearchService_replicaCount (824.73s)
--- PASS: TestAccSearchService_identityUserAssigned (715.61s)
--- PASS: TestAccSearchService_ipRules (807.39s)
--- PASS: TestAccSearchService_complete (811.68s)
--- PASS: TestAccSearchService_update (1728.68s)
```

I was not able to run the following listed 3 tests successfully since they hardcoded create resources in westus which does not work in my setup:

- TestAccSearchService_semanticSearchBasicFreeSkuError
- TestAccSearchService_semanticSearchUpdate
- TestAccSearchService_freeSku

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `search` - update api version to 2024-06-01-preview [GH-27803]

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change

## Related Issue(s)
related to #26575

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: internal/services/search/search_service_resource_test.go [hunk 1]
```diff
@@ -242,6 +242,64 @@ func TestAccSearchService_identity(t *testing.T) {
...
+			Config: r.basic(data, "standard"),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep(),
+	})
+}
+
+func TestAccSearchService_identitySystemAndUserAssigned(t *testing.T) {
```
Comment (jackofallops) [inline internal/services/search/search_service_resource_test.go:274]: Can we change these to being a single create, and then add an `_update` that cycles through the configs?  This then tests the create time inclusion of the identity block of each type, and covers the updating of it. 
so we'd have:
* `TestAccSearchService_identityUserAssigned` - just using `Config: r.identityUserAssigned(data),` and the `data.ImportStep()`
* `TestAccSearchService_identitySystemAndUserAssigned` - just using `r.identitySystemAndUserAssigned(data)` and `data.ImportStep()` 
* a new `TestAccSearchService_identityUpdate` test that runs through a basic -> UserAssigned -> UserAssignedSystemAssigned -> basic lifecycle

**Review by jackofallops — CHANGES_REQUESTED** [review]
Thanks @jan-mrm - This looks mostly good to me, just a change of approach for testing to take a look at please. Once that's addressed, I think this will be good to run though our testing and get merged.

Thanks

File: internal/services/search/search_service_resource_test.go (hunk 1 above)
Comment (jan-mrm) [reply internal/services/search/search_service_resource_test.go]: thanks @jackofallops for your review and feedback!
I updated the tests.
The update test cycles: basic -> identity (which is system assigned) -> user assigned -> system and user assigned -> basic.

One side question: 
Could/Should I have done this api update differently? 
I read I'm supposed to split api updates and new features. After the api update here the type of the returned object changed to I was thinking its ok to do that change as well. The alternative would have been to map / cast back to only system-assigned, which felt cumbersome.

**Review by jackofallops — APPROVED** [review]
Thanks for the changes @jan-mrm - This LGTM now. I'll get the acceptance tests run and assuming there's no problems / regressions we'll get this merged

Thanks

**Comment by jackofallops** [comment]
Tests look good:

image

**Pull request merged**